### PR TITLE
add replace_all documentation

### PR DESCRIPTION
**Pull request by kroening**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/util/string_utils.cpp [hunk 1]
```diff
@@ -162,17 +162,19 @@ std::string escape(const std::string &s)
...
 /// Attributed to Gauthier Boaglio
 /// Source: https://stackoverflow.com/a/24315631/7501486
 /// Used under MIT license
+/// Note that this has quadratic complexity if from.size() != to.size()
 void replace_all(
   std::string &str,
   const std::string &from,
   const std::string &to)
 {
+  PRECONDITION(from.size() != 0);
```
Comment (tautschnig) [inline src/util/string_utils.cpp]: I think `!from.empty()` is more idiomatic?

File: src/util/string_utils.cpp [hunk 2]
```diff
@@ -162,17 +162,19 @@ std::string escape(const std::string &s)
...
 /// Replace all occurrences of a string inside a string
 /// \param [out] str: string to search
-/// \param from: string to replace
+/// \param from: string to replace; must be non-empty
 /// \param to: string to replace with
 /// Copyright notice:
 /// Attributed to Gauthier Boaglio
 /// Source: https://stackoverflow.com/a/24315631/7501486
 /// Used under MIT license
+/// Note that this has quadratic complexity if from.size() != to.size()
```
Comment (tautschnig) [inline src/util/string_utils.cpp]: Quadratic in ... what? The length of `str`? That's equally true then of the implementation that triggered this discussion in #3463.

**Review by tautschnig — CHANGES_REQUESTED** [review]
I don't like semi-formal arguments.

File: src/util/string_utils.cpp (hunk 1 above)
Comment (kroening) [reply src/util/string_utils.cpp]: ok

File: src/util/string_utils.cpp (hunk 2 above)
Comment (kroening) [reply src/util/string_utils.cpp]: Ok, changed wording

**Review by smowton — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 4ecf21186d9585e2c4c3e563570dfda55818250c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/92976712
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 366d56385eaf25ba802dc007ad373c88162447a1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/92977248

File: src/util/string_utils.cpp [hunk 3]
```diff
@@ -162,17 +162,21 @@ std::string escape(const std::string &s)
...
-/// \param from: string to replace
+/// \param from: string to replace; must be non-empty
 /// \param to: string to replace with
 /// Copyright notice:
 /// Attributed to Gauthier Boaglio
 /// Source: https://stackoverflow.com/a/24315631/7501486
 /// Used under MIT license
+/// Note that this is quadratic in str.size() if from.size() != to.size().
+/// There is no guarantee that it is linear in str.size() if
+/// from.size() == to.size().
```
Comment (tautschnig) [inline src/util/string_utils.cpp:173]: Why not?

File: src/util/string_utils.cpp [hunk 4]
```diff
@@ -162,17 +162,21 @@ std::string escape(const std::string &s)
...
 /// Replace all occurrences of a string inside a string
 /// \param [out] str: string to search
-/// \param from: string to replace
+/// \param from: string to replace; must be non-empty
 /// \param to: string to replace with
 /// Copyright notice:
 /// Attributed to Gauthier Boaglio
 /// Source: https://stackoverflow.com/a/24315631/7501486
 /// Used under MIT license
+/// Note that this is quadratic in str.size() if from.size() != to.size().
```
Comment (tautschnig) [inline src/util/string_utils.cpp:171]: ... , because each replacement entails a copy of the contents of the string.

Or maybe something else?

**Comment by tautschnig** [comment]
The precondition is important, but instead of the documentation dance about complexity we might just fix the implementation?
````c++
str result;
std::string::size_type cursor = 0;
for(std::string::size_type replace_at = str.find(from, 0); replace_at != std::string::npos; replace_at = str.find(from, replace_at))
{
  result += str.substr(cursor, replace_at - cursor);
  result += to;
  cursor += from.length();
  replace_at += from.length();
}
if(cursor != str.size())
  result += str.substr(cursor);
str.swap(result);
````

File: src/util/string_utils.cpp (hunk 4 above)
Comment (kroening) [reply src/util/string_utils.cpp:171]: No, that would not change the complexity in str.size() -- unfortunately, std::string::replace does not provide the guarantee that replacing equal-length strings is not independent of the length of the containing string.

**Comment by kroening** [comment]
There are three cases:
1) replacement by shorter string, 2) by string with same length, and 3) by longer string

On an experiment with strings of length 6000:
There's a visible speedup for case 1), but only 25%, and it really needs a long string to kick in. The new implementation is ca. 3-4x slower for case 2). It's ca. 5x slower for case 3).

Thus, I'd propose to stick to the current one, with a warning that it may be slow for long strings.

**Comment by tautschnig** [comment]
@kroening many thanks for the analysis. What about inserting a `result.reserve(str.length());` - it might also be a matter of how many replacements are to be done (i.e., is time lost in the append operation or in doing `find`)? Admittedly, I am surprised by the huge slow-down.

**Comment by kroening** [comment]
The .reserve() makes hardly any measurable difference (the default enlargement strategy seems to work well).
The example I use replaces 30%.

**Review by NlightNFotis — APPROVED** [review]
LGTM, appreciate the performance related comments as well. Would prefer the (performance) data points to be included if it would be possible, but don't mind it either way.

**Review by tautschnig — APPROVED** [review]
Approving after sitting on it for way too long, but really I think this function should just be removed:
1) We do not have a single use in the code base.
2) It's completely unrelated to program analysis or any other CBMC matter.
3) If anyone else uses it (AFAIK TG does), they are more than welcome to copy the implementation.

Any user can then optimise it for their use case.

**Pull request merged**